crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -125,11 +124,22 @@ pub trait SystemParamFetch<'world, 'state>: SystemParamState {
     unsafe fn get_param(
         state: &'state mut Self,
         system_meta: &SystemMeta,
-        world: &'world World,
+        world: &mut SemiSafeCell<'world, World>,
         change_tick: u32,
     ) -> Self::Item;
 }
 
+/// A [`SystemParamFetch`] type that knows if it mutably references the [`World`].
+///
+/// # Safety
+///
+/// - [`is_exclusive`](MaybeExclusive::is_exclusive) must return `true`
+/// if the type's [`get_param`](SystemParamFetch::get_param) constructs a [`&mut World`](World)
+/// or a type that can do so.
+pub unsafe trait MaybeExclusive {
```

Comment:
Why is this trait seperate, instead of the method just being added to `SystemParam` ?

Reply:
Just didn't think to. Would everyone prefer that?

Reply:
Adding to `SystemParam` gets my 'vote'

Reply:
Adding my vote to that.

Reply:
Oh, actually, `SystemParam` isn't an `unsafe` trait.

@DJMcNab, does your suggestion below make it so implementing this method falsely cannot cause UB?

Reply:
A simpler alternative could be to simply make `SystemParam` an unsafe trait too.

Reply:
I'm perfectly on board with making `SystemParam` unsafe. It seems unrealistic for it to remain safe for any extended period of time.

Reply:
After Discord convo, since we want the panics to happen in `<Param::Fetch as SystemParamState>::init`, I moved the method onto `SystemParamState`, which is already an `unsafe` trait.